src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,505 @@
+    layer and updates the weights accordingly.
+
+    Parameters:
+    -----------
+    model: (torch.nn.Module)
+        The model to compute the SVD for.
+    peft_config: (LoraConfig)
+        The configuration for the LoRA layers.
+    dataloader: (torch.utils.data.DataLoader)
+        The dataloader to use for the forward pass.
+    forward_fn: callable
+        The forward function to use for the forward pass. `model(**inputs)` is used if forward_fn is not provided.
+    get_indices_fn: (Optional[callable])
+        The function to use if not all positions in the input tensor should be used for SVD (e.g. for causal language
+        modeling). Can be set to None if all positions should be used. `peft.tuners.lora.eva.get_indices_fn_causal_lm`
+        is used by default Should always return a tensor of shape (n_indices, layer_input.ndim-1).
+    device: (Optional[Union[str, torch.device]])
+        The device to move the model to. If None, the model is not moved.
+    adapter_name: (str)
+        The name of the adapter to initialize the weights for.
+    prepare_activations_fn: (Union[callable, Dict[str, callable], None])
+        If a layer receives multiple inputs as a list, per default the first input is used. This function can be used
+        to modify this behaviour. Accepts a dictionary with layer names as keys. Default logic:
+        ```
+        try:
+            svd_input = activations.detach()
+        except AttributeError:
+            svd_input = activations[0].detach()
+        ```
+
+    Returns:
+    -------- model: (torch.nn.Module)
+        The model with the initialized LoRA weights.
+    """
+    orig_device = get_device_with_meta_params(model)
+    if device is not None:
+        recursive_apply(model, device)
+
+    # assign default eva config
+    if peft_config.eva_config is None:
+        peft_config.eva_config = EvaConfig()
+
+    # compute svd
+    with model.disable_adapter():
+        eva_state_dict = get_eva_state_dict(
+            model=model,
+            peft_config=peft_config,
+            dataloader=dataloader,
+            forward_fn=forward_fn,
+            get_indices_fn=get_indices_fn,
+            prepare_activations_fn=prepare_activations_fn,
+        )
+
+    # assert all lora layers are contained in eva_state_dict
+    missing_eva_inits = []
+    for name, module in model.named_modules():
+        if not isinstance(module, LoraLayer):
+            continue
+        if name in eva_state_dict:
+            w = eva_state_dict.pop(name)
+            module.update_layer(
+                adapter_name,
+                w.size(0),
+                peft_config.lora_alpha,
+                peft_config.lora_dropout,
+                peft_config.init_lora_weights,
+                peft_config.use_rslora,
+                peft_config.use_dora,
+            )
+            module.lora_A[adapter_name].weight.copy_(w)
+        else:
+            module.update_layer(
```

Comment:
What would be a situation when this can occur? During the unit testing, it is not covered.

Reply:
Its mostly because we dont support embedding and conv layers at the moment. So if adapters are added to those layers we cant initialize it with layer so the next best thing would be to do normal lora initalization by default.